source/adapters/level_zero/context.cpp

```diff
@@ -510,6 +511,12 @@ ur_result_t ur_context_handle_t_::getFreeSlotInExistingOrNewPool(
       ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
     if (ProfilingEnabled)
       ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_KERNEL_TIMESTAMP;
+    if (CounterBasedEventEnabled) {
+      ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
+      ze_event_pool_counter_based_exp_desc_t counterBasedExt = {
```

Comment:
I think you might need to update the `getZeEventPoolCache(HostVisible, ProfilingEnabled, ZeDevice);` call above. Otherwise, based on what the queue was that happened to call this function first for this combination of parameters, the event pool might or might not have this flag set.

And, again, I suggest a small refactoring here to use flags (or some other mechanism) instead of passing yet another boolean.

Reply:
The Queue should have the flag set on creation. Because of that, all the subsequent eventpools should be cached with counter-based events enabled.

Reply:
But this cache is in the context, so if you have two queues, one with counter-based events enabled and one without, the one that was first in this function will create the event pool.

Reply:
event pools are stored in context not in a queue
There may be different types of queues coming in and they would require different type of events.
Pool cannot return incompatible event.

Reply:
Using the same mechanism as previously implemented, the eventpool cache returned will be guaranteed to be with counterbased enabled if it is selected, since now there is a portion of the eventpoolcache that is reserved just for counterbased event pools.

Reply:
👍 this desperately needs a refactor though :)

Reply:
I'll file a ticket tomorrow for all the refactoring, should be pretty straight forward 👍